File: java/components/StrataPersistence/src/strata/persistence/inmemory/InMemoryUnitOfWork.java
// ##########################################################################
// # File Name:	InMemoryUnitOfWork.java
// #
// # Copyright:	2012, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataEntity Framework.
// #
// #   			The StrataEntity Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataEntity Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataEntity
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.persistence.inmemory;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import strata.persistence.namedquery.INamedQuery;
import strata.persistence.repository.IKeyRetriever;
import strata.persistence.unitofwork.AbstractUnitOfWork;
import strata.persistence.unitofwork.CommitFailedException;
import strata.persistence.unitofwork.RollbackFailedException;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class InMemoryUnitOfWork
    extends AbstractUnitOfWork
{
    private Map<EntityIdentifier,Object> itsInserted;
    private Map<EntityIdentifier,Object> itsUpdated;
    private Map<EntityIdentifier,Object> itsRemoved;
    private Map<EntityIdentifier,Object> itsEntities;
    
    /************************************************************************
     * Creates a new {@code InMemoryUnitOfWork}. 
     *
     */
    public 
    InMemoryUnitOfWork(
        InMemoryUnitOfWorkProvider   provider,
        Map<EntityIdentifier,Object> entities)
    {
        super( provider );
        itsInserted  = new HashMap<EntityIdentifier,Object>();
        itsUpdated   = new HashMap<EntityIdentifier,Object>();
        itsRemoved   = new HashMap<EntityIdentifier,Object>();
        itsEntities  = entities;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    close() throws Exception
    {
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> E 
    doInsertNew(Class<K> keyType,Class<E> entityType,E newEntity)
    {
        E                insertedEntity = null;
        K                key            = null;
        EntityIdentifier id             = null;
        
        insertedEntity = 
            getReplicator(keyType,entityType).replicate(newEntity);
        key =             
            getRetriever(keyType,entityType).getKey(insertedEntity);
        id = new EntityIdentifier(entityType,key);
        
        if ( itsUpdated.containsKey( id ) )
            throw new IllegalArgumentException("entity is already updated");
        
        if ( itsRemoved.containsKey( id ) )
            throw new IllegalArgumentException("entity is already removed");
        
        if ( itsInserted.containsKey( id ) || doHasExisting( entityType,key ) )
            throw new IllegalArgumentException("entity is already inserted");
        
        itsInserted.put( id,insertedEntity );
        return insertedEntity;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> E 
    doUpdateExisting(Class<K> keyType,Class<E> entityType,E existingEntity)
    {
        E                updatedEntity = null;
        K                key           = null;
        EntityIdentifier id            = null;
        
        updatedEntity = 
            getReplicator(keyType,entityType).replicate(existingEntity);
        key =             
            getRetriever(keyType,entityType).getKey(updatedEntity);
        id = new EntityIdentifier(entityType,key);
        
        if ( itsRemoved.containsKey( id ) )
            throw new IllegalArgumentException("entity is already removed");

        if ( !itsInserted.containsKey(id) && !doHasExisting(entityType,key))
            throw new IllegalArgumentException("entity does not exist");
        
        itsUpdated.put( new EntityIdentifier(entityType,key),updatedEntity );
        return updatedEntity;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> void 
    doRemoveExisting(Class<K> keyType,Class<E> entityType,E existingEntity)
    {
        K                key = null;
        EntityIdentifier id  = null;
        
        key =             
            getRetriever(keyType,entityType).getKey(existingEntity);
        id = new EntityIdentifier(entityType,key);

        if ( !itsInserted.containsKey( id ) && !doHasExisting( entityType,key ) )
            throw new IllegalArgumentException("entity does not exist");
        
        itsRemoved.put( id,existingEntity );        
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> E 
    doGetExisting(Class<E> type,K key)
    {
        EntityIdentifier id = new EntityIdentifier(type,key);
        
        if ( itsRemoved.containsKey( id ) )
            return null;
        
        if ( itsUpdated.containsKey( id ) )
            return type.cast( itsUpdated.get( id ) );
        
        if ( itsInserted.containsKey( id ) )
            return type.cast( itsInserted.get( id ) );
        
        if ( itsEntities.containsKey( id ) )
            return type.cast( itsEntities.get( id ) );
        
        return getCandidates( key.getClass(),type ).get( key );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <E> INamedQuery<E> 
    doGetNamedQuery(Class<E> type,String queryName)
    {
        return 
            ((InMemoryUnitOfWorkProvider)getProvider())
                .getNamedQuery( type,queryName );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <K extends Serializable,E> boolean 
    doHasExisting(Class<E> type,K key)
    {
        EntityIdentifier id = new EntityIdentifier(type,key);
        
        if ( itsRemoved.containsKey( id ) )
            return false;
        
        if ( itsUpdated.containsKey( id ) )
            return true;
        
        if ( itsInserted.containsKey( id ) )
            return true;
        
        if ( itsEntities.containsKey( id ) )
            return true;
        
        return getCandidates( key.getClass(),type ).containsKey( key );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected <E> boolean 
    doHasNamedQuery(Class<E> type,String queryName)
    {
        return doGetNamedQuery( type,queryName ) != null;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected void 
    doCommit()
        throws CommitFailedException
    {
        
        try
        {
            commitInserted();
            commitUpdated();
            commitRemoved();
        }
        catch (Exception cause)
        {
            throw new CommitFailedException(cause);
        }
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    protected void 
    doRollback()
        throws RollbackFailedException
    {
        itsInserted.clear();
        itsUpdated.clear();
        itsRemoved.clear();
    }


    /************************************************************************
     *  
     *
     */
    protected void
    commitInserted()
    {
        for (
            Map.Entry<EntityIdentifier,Object> entry : 
                itsInserted.entrySet())
            itsEntities.put( entry.getKey(),entry.getValue() );
    }
    
    /************************************************************************
     *  
     *
     */
    protected void
    commitUpdated()
    {
        for (
            Map.Entry<EntityIdentifier,Object> entry : 
                itsUpdated.entrySet())
            itsEntities.put( entry.getKey(),entry.getValue() );
    }
   
    /************************************************************************
     *  
     *
     */
    protected void
    commitRemoved()
    {
        for (
            Map.Entry<EntityIdentifier,Object> entry : 
                itsRemoved.entrySet())
        {
            if ( itsEntities.remove( entry.getKey() ) == null )
                throw new IllegalStateException("entity does not exist");
        }
    }

    /************************************************************************
     *  
     *
     * @param keyType
     * @param entityType
     * @return
     */
    private <K extends Serializable,E> IEntityReplicator<K,E> 
    getReplicator(Class<K> keyType,Class<E> entityType)
    {
        return 
            ((InMemoryUnitOfWorkProvider)getProvider())
                .getReplicator( keyType,entityType );
    }

    /************************************************************************
     *  
     *
     * @param keyType
     * @param entityType
     * @return
     */
    private <K extends Serializable,E> IKeyRetriever<K,E> 
    getRetriever(Class<K> keyType,Class<E> entityType)
    {
        return 
            ((InMemoryUnitOfWorkProvider)getProvider())
                .getRetriever( keyType,entityType );
    }

    /************************************************************************
     *  
     *
     * @param keyType
     * @param entityType
     * @return
     */
    private <K,T> Map<K,T>
    getCandidates(Class<K> keyType,Class<T> entityType)
    {
        Map<K,T> candidates = new HashMap<K,T>();
        
        for (
            Map.Entry<EntityIdentifier,Object> entry:
                itsEntities.entrySet())
        {
            if ( entityType.isAssignableFrom( entry.getKey().getType() ) )
            {
                try
                {
                    candidates.put( 
                        keyType.cast( entry.getKey().getKey() ),
                        entityType.cast( entry.getValue() ) );
                }
                catch (ClassCastException e) {}
            }
        }
        
        return candidates;
    }
}

// ##########################################################################

File: legacy/csharp/components/Strata.Persistence/InMemory/InMemoryUnitOfWork.cs
//  ##########################################################################
//  # File Name: InMemoryUnitOfWork.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using Strata.Foundation.Utility;
using Strata.Persistence.NamedQuery;
using Strata.Persistence.UnitOfWork;

namespace Strata.Persistence.InMemory
{

    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    ///  
    public
    class InMemoryUnitOfWork:
        AbstractUnitOfWork
    {
        public bool IsDisposed { get; protected set; }

        private readonly IDictionary<EntityIdentifier,Object>  inserted;
        private readonly IDictionary<EntityIdentifier,Object>  updated;
        private readonly IDictionary<EntityIdentifier,Object>  removed;
        private readonly IDictionary<EntityIdentifier,Object>  entities;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>InMemoryUnitOfWork</c> instance.
        /// </summary>
        /// 
        /// <param name="ent">in memory entities</param>
        /// 
        public 
        InMemoryUnitOfWork(
            InMemoryUnitOfWorkProvider           provider,
            IDictionary<EntityIdentifier,Object> ent):
            base(provider)
        {
            inserted   = new Dictionary<EntityIdentifier,Object>();
            updated    = new Dictionary<EntityIdentifier,Object>();
            removed    = new Dictionary<EntityIdentifier,Object>();
            entities   = ent;
            IsDisposed = false;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void
        Dispose()
        {
            inserted.Clear();
            updated.Clear();
            removed.Clear();
            IsDisposed = true;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override T 
        DoInsert<K,T>(T entity)
        {
            T                insertedEntity = null;
            K                key            = default(K);
            EntityIdentifier id             = null;
        
            insertedEntity = GetReplicator<T>().Replicate(entity);
            key            = GetRetriever<K,T>()(insertedEntity);
            id             = new EntityIdentifier(typeof(T),key);
        
            if ( updated.ContainsKey( id ) )
                throw new ArgumentException("entity is already updated");
        
            if ( removed.ContainsKey( id ) )
                throw new ArgumentException("entity is already removed");
        
            if ( inserted.ContainsKey(id) || DoHasUniqueWithKey<K,T>(key))
                throw new ArgumentException("entity is already inserted");
        
            inserted.Add( id,insertedEntity );
            return insertedEntity;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override T 
        DoUpdate<K,T>(T entity)
        {
            T                updatedEntity = null;
            K                key           = default(K);
            EntityIdentifier id            = null;
        
            updatedEntity = GetReplicator<T>().Replicate(entity);
            key =  GetRetriever<K,T>()(updatedEntity);
            id = new EntityIdentifier(typeof(T),key);
        
            if ( removed.ContainsKey( id ) )
                throw new ArgumentException("entity is already removed");

            if ( !inserted.ContainsKey(id) && !DoHasUniqueWithKey<K,T>(key))
                throw new ArgumentException("entity does not exist");
        
            updated.Add( id,updatedEntity );
            return updatedEntity;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void 
        DoRemove<K,T>(T entity)
        {
            K                key = default(K);
            EntityIdentifier id  = null;
        
            key = GetRetriever<K,T>()(entity);
            id = new EntityIdentifier(typeof(T),key);

            if ( !inserted.ContainsKey(id) && !DoHasUniqueWithKey<K,T>(key))
                throw new ArgumentException("entity does not exist");
        
            removed.Add( id,entity );        
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override T 
        DoGetUniqueWithKey<K,T>(K key)
        {
            EntityIdentifier id = new EntityIdentifier(typeof(T),key);
        
            if ( removed.ContainsKey( id ) )
                return null;
 
            if ( updated.ContainsKey( id ) )
                return (T)updated[id];

            if ( inserted.ContainsKey( id ) )
                return (T)inserted[id];
        
            return GetReplicator<T>().Replicate(GetCandidates<K,T>()[key]);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override T 
        DoGetUniqueWithPredicate<T>(Expression<Func<T,bool>> predicate)
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier,Object>(entities);
            T result =
                queryable
                    .Values
                    .OfType<T>()
                    .AsQueryable()
                    .Where(predicate)
                    .SingleOrDefault();

            if (result != null)
                return GetReplicator<T>().Replicate(result);

            queryable = new Dictionary<EntityIdentifier,object>();
            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

            return
                queryable
                    .Values
                    .OfType<T>()
                    .AsQueryable()
                    .Where(predicate)
                    .SingleOrDefault();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override IList<T> 
        DoGetAllWithPredicate<T>(Expression<Func<T,bool>> predicate)
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier,Object>(entities);
            IList<T> temp =
                queryable
                    .Values
                    .OfType<T>()
                    .AsQueryable()
                    .Where(predicate)
                    .ToList();
            IList<T> results = new List<T>();

            if (temp != null && temp.Count > 0)
                foreach (T entity in temp)
                    results.Add(GetReplicator<T>().Replicate(entity));

            queryable = new Dictionary<EntityIdentifier,object>();
            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

            results.AddAll(
                queryable
                    .Values
                    .OfType<T>()
                    .AsQueryable()
                    .Where(predicate)
                    .ToList());

            return results;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override IList<T> 
        DoGetAll<T>()
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier,Object>(entities);
            IList<T> temp =
                queryable
                    .Values
                    .OfType<T>()
                    .ToList();
            IList<T> results = new List<T>();

            if (temp != null && temp.Count > 0)
                foreach (T entity in temp)
                    results.Add(GetReplicator<T>().Replicate(entity));

            queryable = new Dictionary<EntityIdentifier,object>();
            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

            results.AddAll(
                queryable
                    .Values
                    .OfType<T>()
                    .ToList());

            return results;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override INamedQuery<T> 
        DoGetNamedQuery<T>(string queryName)
        {
            return
                ((InMemoryUnitOfWorkProvider) Provider)
                    .GetNamedQuery<T>(queryName);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override bool 
        DoHasUniqueWithKey<K,T>(K key)
        {
            EntityIdentifier id = new EntityIdentifier(typeof(T),key);
        
            if ( removed.ContainsKey( id ) )
                return false;
 
            if ( updated.ContainsKey( id ) )
                return true;

            if ( inserted.ContainsKey( id ) )
                return true;
        
            return GetCandidates<K,T>().ContainsKey( key );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override bool 
        DoHasAnyWithPredicate<T>(Expression<Func<T,bool>> predicate)
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier,Object>(entities);

            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

            return
                queryable
                    .Values
                    .OfType<T>()
                    .AsQueryable()
                    .Where(predicate)
                    .Any();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override bool 
        DoHasNamedQuery<T>(string queryName)
        {
            return ((InMemoryUnitOfWorkProvider) Provider).HasNamedQuery<T>(queryName);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void 
        DoCommit()
        {
            try
            {
                CommitInserted();
                CommitUpdated();
                CommitRemoved();

                inserted.Clear();
                updated.Clear();
                removed.Clear();            
            }
            catch (Exception cause)
            {
                throw new CommitFailedException("commit failed",cause);
            }      
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void 
        DoRollback()
        {
            try
            {
                inserted.Clear();
                updated.Clear();
                removed.Clear();
                base.DoRollback();
            }
            catch (RollbackFailedException)
            {
                throw;
            }
            catch (Exception cause)
            {
                throw new RollbackFailedException("rollback failed",cause);
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public T 
        DoInsert<K,T>(
            T                    entity,
            IEntityReplicator<T> replicator,
            KeyRetriever<K,T>    retriever)
            where T:class
        {
            T                insertedEntity = replicator.Replicate( entity );
            K                key = retriever( insertedEntity );
            EntityIdentifier id  = new EntityIdentifier(typeof(T),key);   
     
            if ( updated.ContainsKey( id ) )
                throw new InvalidOperationException("entity is already updated");
        
            if ( removed.ContainsKey( id ) )
                throw new InvalidOperationException("entity is already removed");
        
            if ( inserted.ContainsKey( id ) || DoHas<K,T>( key ) )
                throw new InvalidOperationException("entity is already inserted");
        
            inserted.Add( id,insertedEntity );
            return insertedEntity;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public  T 
        DoUpdate<K,T>(
            T                    entity,
            IEntityReplicator<T> replicator,
            KeyRetriever<K,T>    retriever)
            where T:class
        {
            T                updatedEntity = replicator.Replicate( entity );
            K                key = retriever( updatedEntity );
            EntityIdentifier id  = new EntityIdentifier(typeof(T),key);   
        
            if ( !inserted.ContainsKey( id ) && !DoHas<K,T>( key ) )
                throw new InvalidOperationException("entity does not exist");
        
            if ( removed.ContainsKey( id ) )
                throw new InvalidOperationException("entity is already removed");
        
            if ( updated.ContainsKey( id ) )
                updated[id] = updatedEntity;
            else
                updated.Add( id,updatedEntity );

            return updatedEntity;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public void 
        DoRemove<K,T>(T entity,KeyRetriever<K,T> retriever)
        {
            K                key = retriever( entity );
            EntityIdentifier id  = new EntityIdentifier( typeof(T),key );       
         
            if ( !inserted.ContainsKey( id ) && !DoHas<K,T>( key ) )
                throw new InvalidOperationException("entity does not exist");

            removed.Add( id,entity );
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public T
        DoGet<K,T>(K key)
            where T:class
        {
            EntityIdentifier id = new EntityIdentifier( typeof(T),key );

            if ( removed.ContainsKey( id ) )
                return null;

            if ( updated.ContainsKey( id ) )
                return (T)updated[id];

            if ( inserted.ContainsKey( id ) )
                return (T)inserted[id];

            return (T)entities[id];
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public IQueryable<T>
        DoQuery<T>()
        {
            IDictionary<EntityIdentifier,Object> queryable = 
                new Dictionary<EntityIdentifier, Object>(entities);

            queryable = queryable.Merge( inserted );
            queryable = queryable.Merge( updated );
            queryable = queryable.Filter( removed );

             return
                queryable
                    .Values
                    .OfType<T>()
                    .AsQueryable();
        }
    
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public bool
        DoHas<K,T>(K key)
        {
            EntityIdentifier id = new EntityIdentifier( typeof(T),key );

            if ( removed.ContainsKey( id ) )
                return false;

            if ( inserted.ContainsKey( id ) )
                return true;

            if ( updated.ContainsKey( id ) )
                return true;

            return entities.ContainsKey( id );
        }
    
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public void
        CommitInserted()
        {
            foreach (
                KeyValuePair<EntityIdentifier,Object> entry in inserted)
                entities.Add( entry.Key,Replicate(entry.Value) );
        }
    
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public void
        CommitUpdated()
        {
            foreach (
                KeyValuePair<EntityIdentifier,Object> entry in updated)
                entities[entry.Key] = Replicate(entry.Value);
        }
   
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public void
        CommitRemoved()
        {
            foreach (
                KeyValuePair<EntityIdentifier,Object> entry in removed)
            {
                if ( entities.Remove( entry.Key ) == false )
                    throw 
                        new InvalidOperationException("entity does not exist");
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public IEntityReplicator<T>
        GetReplicator<T>()
            where T: class
        {
            return 
                ((InMemoryUnitOfWorkProvider)Provider).GetReplicator<T>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public KeyRetriever<K,T>
        GetRetriever<K,T>()
            where T: class
        {
            return 
                ((InMemoryUnitOfWorkProvider)Provider).GetRetriever<K,T>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private IDictionary<K,T>
        GetCandidates<K,T>()
            where T: class
        {
            IDictionary<K,T> candidates = new Dictionary<K,T>();
            KeyRetriever<K,T> retriever = GetRetriever<K,T>();

            foreach (T entity in GetEntitiesByType<T>())
                candidates.Add(retriever(entity),entity);
                
            return candidates;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private IList<T>
        GetEntitiesByType<T>()
        {
            Type     type   = typeof(T);
            IList<T> output = new List<T>();
        
            foreach (
                KeyValuePair<EntityIdentifier,Object> entry 
                    in entities)
            {
                if ( type.IsAssignableFrom( entry.Key.EntityType ) )
                    output.Add( (T)( entry.Value ) );
            }
        
            return output;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private object
        Replicate(object source)
        {
            IEntityReplicator<object> replicator = 
                ((InMemoryUnitOfWorkProvider)Provider).GetObjectReplicator(
                    source.GetType());

            return replicator.Replicate(source);
        }
    }
}

//  ##########################################################################
